### ISSUE TITLE
False positives on UnsafeCast on AppVeyor (Windows?) only

### ISSUE THREAD

**Issue by rock3r** [arturbosch/detekt#1137]
## Steps to Reproduce (for bugs)
See https://github.com/arturbosch/detekt/blob/fab854abc1d1efdc6218dc8e5cf486f6fb194de2/detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/ThresholdedCodeSmellAssert.kt

On *nix (my macOS machine, and on Travis' CI machines) I get no warnings.

On Windows (AppVeyor) we get these warnings:

```
Ruleset: potential-bugs - 1h debt
	UnsafeCast - [isThresholded] at C:\projects\detekt\detekt-test\src\main\kotlin\io\gitlab\arturbosch\detekt\test\ThresholdedCodeSmellAssert.kt:12:36
	UnsafeCast - [smell] at C:\projects\detekt\detekt-test\src\main\kotlin\io\gitlab\arturbosch\detekt\test\ThresholdedCodeSmellAssert.kt:26:15
	UnsafeCast - [smell] at C:\projects\detekt\detekt-test\src\main\kotlin\io\gitlab\arturbosch\detekt\test\ThresholdedCodeSmellAssert.kt:35:15
```

The warnings on Windows are incorrect, as those casts are all guarded behind an assertion. Neither the compiler nor the IDE complain about those.

## Your Environment
* Version used: RC9
* Operating System and version: macOS 10.13.6/Travis CI/AppVeyor
* Link to your project: Detekt :)

**Comment by 3flex** [comment]
This has come up previously: https://github.com/arturbosch/detekt/issues/926

**Comment by rock3r** [comment]
Hm I missed that, sorry. In this case though those are plain false positives...

**Comment by 3flex** [comment]
Oh sorry, I meant more that the difference in results between *nix and Windows had come up previously (but again with UnsafeCast).

**Comment by rock3r** [comment]
Ah yeah no problem mate :) I don't know what makes that rule so platform-reliant... Haven't had time to look at the implementation. I plan on doing, but I really have little time these days

**Comment by arturbosch** [comment]
The implementation just uses a function provided by the kotlin compiler, so I guess we need to track this down on the kotlin side.

**Comment by 3flex** [comment]
I've looked into this more closely - the current check provided by detekt does not respect smart casts. The compiler's implementation only checks the format of the expression and looks for the `as` keyword.

To fix this so the check respects smart casts I expect we need type and symbol resolution.

Didn't look into differences between Windows & Linux.
